### ISSUE TITLE
Invalid `try` operand type deduction

### ISSUE THREAD

**Issue by amp-59** [ziglang/zig#21296]
### Zig Version

0.14.0-dev.1411+a670f5519

### Steps to Reproduce and Observed Behavior

Compile example program with `zig build-obj try_operand_ty_error.zig`
`try_operand_ty_error.zig`:
```zig
var eu: error{}!void = {};

fn fn0() error{}!void {
    return @as(void, try eu); // 0
}
fn fn1() void {
    return @as(void, try eu); // 1
}
fn fn2() void {
    return @as(error{a}!void, try eu); // 2
}

comptime {
    _ = &fn0;
}
comptime {
    _ = &fn1;
}
comptime {
    _ = &fn2;
}
```

Output:
```
zig build-obj try_operand_ty_error.zig
try_operand_ty_error.zig:7:22: error: expected 'void', found error set
    return @as(void, try eu); // 1
                     ^~~~~~
try_operand_ty_error.zig:6:10: note: function cannot return an error
fn fn1() void {
         ^~~~
try_operand_ty_error.zig:10:31: error: expected 'void', found error set
    return @as(error{a}!void, try eu); // 2
                              ^~~~~~
try_operand_ty_error.zig:9:10: note: function cannot return an error
fn fn2() void {
         ^~~~
```

### Expected Behavior

The type system for `try` operands in `return` expressions should be consistent with the rest of the language.

Invalid return types should be determined by coercion, which also gives a better error message.

For example compile the program again with Zig version `0.14.0-dev.1391+e084c46ed`:
```
zig build-obj try_operand_ty_error.zig
try_operand_ty_error.zig:10:12: error: expected type 'void', found 'error{a}!void'
    return @as(error{a}!void, try eu); // 2
           ^~~~~~~~~~~~~~~~~~~~~~~~~~
try_operand_ty_error.zig:10:12: note: cannot convert error union to payload type
try_operand_ty_error.zig:10:12: note: consider using 'try', 'catch', or 'if'
try_operand_ty_error.zig:9:10: note: function cannot return an error
fn fn2() void {
         ^~~~
```

**Comment by mlugg** [comment]
Will be resolved by #21420 (#21414 was a duplicate of this).

**Comment by mlugg** [comment]
Resolved in 19924ca2890964b411362c423dd9f4b10596a18f.
